nexus/codec.go

```diff
@@ -0,0 +1,204 @@
+// as [NewClient] and [NewHTTPHandler].
+// By default, the SDK supports serialization of JSONables, byte slices, and nils.
+type Codec interface {
+	// Encode encodes a value into a [Stream].
+	Encode(any) (*Stream, error)
+	// Decode decodes a [Stream] into a given reference.
+	Decode(*Stream, any) error
+}
+
+type codecChain []Codec
+
+var errCodecIncompatible = errors.New("incompatible codec")
+
+func (c codecChain) Encode(v any) (*Stream, error) {
+	for _, l := range c {
+		p, err := l.Encode(v)
+		if err != nil {
+			if errors.Is(err, errCodecIncompatible) {
+				continue
+			}
+			return nil, err
+		}
+		return p, nil
+	}
+	return nil, errCodecIncompatible
+}
+
+func (c codecChain) Decode(s *Stream, v any) error {
+	lenc := len(c)
+	for i := range c {
+		l := c[lenc-i-1]
+		if err := l.Decode(s, v); err != nil {
+			if errors.Is(err, errCodecIncompatible) {
+				continue
+			}
+			return err
+		}
+		return nil
+	}
+	return errCodecIncompatible
+}
+
+var _ Codec = codecChain{}
+
+type jsonCodec struct{}
+
+func (jsonCodec) Decode(s *Stream, v any) error {
+	if !isMediaTypeJSON(s.Header["content-type"]) {
+		return errCodecIncompatible
+	}
+	body, err := io.ReadAll(s.Reader)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(body, &v)
+}
+
+func (jsonCodec) Encode(v any) (*Stream, error) {
+	body, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	return &Stream{
+		Header: map[string]string{
+			"content-type":   "application/json",
+			"content-length": fmt.Sprintf("%d", len(body)),
+		},
+		Reader: io.NopCloser(bytes.NewReader(body)),
+	}, nil
+}
+
+var _ Codec = jsonCodec{}
+
+type nilCodec struct{}
+
+func (nilCodec) Decode(s *Stream, v any) error {
+	if s.Header["content-length"] != "0" {
+		return errCodecIncompatible
+	}
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Pointer {
+		return fmt.Errorf("could not decode non pointer value: %v", v)
+	}
+	re := rv.Elem()
+	if !re.CanSet() {
+		return fmt.Errorf("non settable type: %v", v)
+	}
+	// Set the zero value for the given type.
+	re.Set(reflect.Zero(re.Type()))
+
+	return nil
+}
+
+func (nilCodec) Encode(v any) (*Stream, error) {
+	if v != nil {
+		rv := reflect.ValueOf(v)
+		if !(rv.Kind() == reflect.Ptr && rv.IsNil()) {
+			return nil, errCodecIncompatible
+		}
+	}
+	return &Stream{
+		Header: map[string]string{"content-length": "0"},
+		Reader: nil,
+	}, nil
+}
+
+var _ Codec = nilCodec{}
+
+type byteSliceCodec struct{}
+
+func (byteSliceCodec) Decode(s *Stream, v any) error {
+	if !isMediaTypeOctetStream(s.Header["content-type"]) {
+		return errCodecIncompatible
+	}
+	if bPtr, ok := v.(*[]byte); ok {
+		if bPtr == nil {
+			return fmt.Errorf("cannot decode into nil pointer: %v", v)
+		}
+		b, err := io.ReadAll(s.Reader)
+		if err != nil {
+			return err
+		}
+		*bPtr = b
+		return nil
+	}
+	// v is *any
+	rv := reflect.ValueOf(v).Elem()
+	if rv.Kind() == reflect.Interface {
```

Comment:
I am a bit confused here on what you're trying to do. The code I think is saying every pointer to an interface you try to set as a byte array. You need to do exact equality check `Type()` with the reflect type of `any`, not any kind of interface.

But I am not convinced we should support `any` for this data type.

Reply:
`reflect.Interface` is `interface{}` (`any`).

I initially didn't support it but saw that Temporal does so I ported that logic.

It kinda makes sense that if you can populate an any, you would.

Reply:
No, `reflect.Interface` is a value of the `Kind` enumerate which represents a _kind_ of a type, not the type. What most would do here is:

```
// store this somewhere global
var anyType := reflect.TypeOf((*any)(nil)).Elem()
rv.Type() == anyType
```

Reply:
Hmm... I thought I saw somewhere that `Interface` kind is `interface{}`. Thanks!

BTW: we should fix this in Temporal, it checks that the `Elem().Kind() == reflect.Interface` which is incorrect: https://github.com/temporalio/sdk-go/blob/0352634b45b1b81293c9bf3cec3bb033b82f841f/converter/byte_slice_payload_converter.go#L61

I think we should keep support for `[]byte -> any`, it'd be up to the user to cast.